.github/workflows/kernelctf-submission-verification.yaml

```diff
@@ -208,86 +240,241 @@ jobs:
+            
+            echo "Downloading and extracting Android system images for $RELEASE_ID (build: $BUILD_NUMBER)..."
+            
+            # Download, extract, and delete phone image immediately
+            wget --progress=dot:giga https://storage.googleapis.com/kernelctf-build/releases/$RELEASE_ID/aosp_cf_x86_64_only_phone-img-${BUILD_NUMBER}.zip 2>&1 | tail -n 1
+            echo "Extracting phone image..."
+            unzip -q aosp_cf_x86_64_only_phone-img-${BUILD_NUMBER}.zip
+            echo "Removing phone image archive to save space..."
+            rm -f aosp_cf_x86_64_only_phone-img-${BUILD_NUMBER}.zip
+            
+            # Download, extract, and delete host package immediately
+            wget --progress=dot:giga https://storage.googleapis.com/kernelctf-build/releases/$RELEASE_ID/cvd-host_package.tar.gz 2>&1 | tail -n 1
+            echo "Extracting host package..."
+            tar -xzf cvd-host_package.tar.gz
+            echo "Removing host package archive to save space..."
+            rm -f cvd-host_package.tar.gz
+            
+            echo "System images ready"
+          ) &
+          DOWNLOAD_PID=$!
+          
+          # Wait for both to complete
+          echo "Running in parallel: dependencies installation (PID: $DEPS_PID) and image download (PID: $DOWNLOAD_PID)..."
+          wait $DEPS_PID
+          DEPS_EXIT=$?
+          wait $DOWNLOAD_PID
+          DOWNLOAD_EXIT=$?
+          
+          # Check if either failed
+          if [ $DEPS_EXIT -ne 0 ]; then
+            echo "❌ Error: Dependencies installation failed with exit code $DEPS_EXIT"
+            echo "Last 50 lines of android_deps_install.log:"
+            tail -n 50 android_deps_install.log || true
+            exit $DEPS_EXIT
+          fi
+          
+          if [ $DOWNLOAD_EXIT -ne 0 ]; then
+            echo "❌ Error: Image download failed with exit code $DOWNLOAD_EXIT"
+            exit $DOWNLOAD_EXIT
+          fi
+          
+          echo "✓ Both tasks completed successfully"
+          
+          # Final cleanup
+          echo "Final cleanup: removing installation log..."
+          rm -f android_deps_install.log 2>/dev/null || true
+
+      - name: Set paths - Android only
+        if: startsWith(matrix.target, 'android')
+        run: |
+          APK_PATH="../../android_deps/kernelctf/server/android_shellserver/app/build/outputs/apk/release/app-release.apk"
+          echo "APK_PATH=$(realpath $APK_PATH)" >> $GITHUB_ENV
+          echo "[INFO] APK path set to: $(realpath $APK_PATH)"
+          
+          echo "ANDROID_RELEASE_PATH=$(realpath android_release)" >> $GITHUB_ENV
 
-      # ugly hack to make Github Actions UI to show repro logs separately in somewhat readable fashion
       - id: repro1
         name: Reproduction (1 / 10)
         continue-on-error: true
-        run: ./repro.sh 1
+        run: |
+          if [[ "$RELEASE_ID" == android* ]]; then
```

Comment:
Move this logic to repro.sh, do not repeat the same logic multiple times, otherwise we have to replace everywhere if anything changes.